### ISSUE TITLE
Cannot use a non-enumerable PropertySource as a Bootstrap PropertySource

### ISSUE THREAD

**Issue by dzou** [spring-cloud/spring-cloud-commons#724]
Hello!

Our project is currently trying to register a custom bootstrap property source that will load properties into our application environment at the bootstrap phase. This has worked great for us.

However, we've recently been trying to register a non-enumerable property source but keep receiving the following error; I believe there is an issue that makes it not possible to register non-enumerable property sources as bootstrap properties.

**Sample**
Reproduction is here: https://github.com/dzou/non-enumerable-prop-source-ex

Error message:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

Failed to bind properties under 'spring.cloud.config' to org.springframework.cloud.bootstrap.config.PropertySourceBootstrapProperties:

    Property: spring.cloud.config
    Value: constant-property
    Origin: "spring.cloud.config" from property source "bootstrapProperties-My-Custom-Bootstrap-Property-Source"
    Reason: Failed to enumerate property names due to non-enumerable property source: CustomPropertySource {name='My-Custom-Bootstrap-Property-Source'}

Action:

Update your application's configuration
```

I traced the issue through the code, I think I know where the source of the issue is.

Basically there are a few interactions:

1. All Bootstrap property sources are initialized and then wrapped in the [BootstrapPropertySource class in this line](https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java#L103).

2. The [BootstrapPropertySource class extends from EnumerablePropertySource](https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/BootstrapPropertySource.java#L32).

3. Then, when all the property sources are checked for whether they are enumerable or not, the framework believes it is enumerable because later the code checks for whether the [property source is instance of EnumerablePropertySource](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/context/properties/source/SpringConfigurationPropertySource.java#L174).

    This is not quite right because in the case of the `BootstrapPropertySource` it should instead check if the delegate is `instanceof EnumerablePropertySource` instead of the property source itself.

Let me know your thoughts on this! We were excited to discover bootstrap properties and very excited to see if it can work for non-enumerable property sources.

cc/ @elefeint @dmitry-s @meltsufin

**Comment by ryanjbaxter** [comment]
Is that exception coming from https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/config/BootstrapPropertySource.java#L51?

**Comment by dzou** [comment]
Yup, the exception is coming from there. Spring framework code appears to be calling that method at some point.

I think the code is fine, but if you look at the #3 point in my post, the Spring code thinks that it is enumerable because it checks if that class is `instance of EnumerablePropertySource` (which it is) but in the case of BootstrapPropertySource, I think it needs to check if the `delegate` field in it is enumerable rather than just check the property source itself.

See also this line where the enumerable check is being made: https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/context/properties/source/SpringConfigurationPropertySource.java#L144

**Comment by spencergibb** [comment]
I mean, the only thing we could do is move the exception into the constructor so it happens earlier. That won't fix your problem on a non-`EnumerablePropertySource`. I'm not sure what you'd have us do.

**Comment by spencergibb** [comment]
What does the non-enumerable property source look like?

**Comment by spencergibb** [comment]
I guess the other option is to have two `BootstrapPropertySource` classes, one enumerable one not.

**Comment by meltsufin** [comment]
We are creating a property source for GCP Secret Manager. It should be non-enumerable because the best practice is to only access the properties (secrets) that are used by the application, not every single secret in the project.

**Comment by spencergibb** [comment]
I wondered about that. I think this is probably the way to go:

> I guess the other option is to have two BootstrapPropertySource classes, one enumerable one not.

Does that make sense?

**Comment by dzou** [comment]
> I wondered about that. I think this is probably the way to go:
> 
> > I guess the other option is to have two BootstrapPropertySource classes, one enumerable one not.
> 
> Does that make sense?

Yeah that sounds good to me.

By the way, our current implementation of our property source we've just been [implementing EnumerablePropertySource but just returning `new String[0]` when `getPropertyNames()` is called](https://github.com/spring-cloud/spring-cloud-gcp/pull/2302/files#diff-c81f0febaa4dbfee981a654aae304b5cR34). It seems to still work, but not sure if there's any negative consequences for doing this.

**Comment by spencergibb** [comment]
hehe, probably harmless, but you should be able to remove that once this is in place.
